keras/src/trainers/compile_utils.py

```diff
@@ -430,202 +429,179 @@ def __init__(
+                    raise KeyError(
+                        f"There are keys: {list(loss_keys - y_true_keys)} in "
+                        "the `loss` argument, but they can't be find found in "
+                        "`y` (`y_true`)."
                     )
-                for name in loss_weights.keys():
-                    if name not in output_names:
-                        raise ValueError(
-                            "In the dict argument `loss_weights`, key "
-                            f"'{name}' does not correspond to any model "
-                            f"output. Received: loss_weights={loss_weights}"
-                        )
-                    if not isinstance(loss_weights[name], float):
-                        raise ValueError(
-                            "For a model with multiple outputs, "
-                            "when providing the `loss_weights` argument as a "
-                            "dict, each dict entry should be a Python float "
-                            "(the weighting coefficient corresponding to the "
-                            f"loss for that output). At key '{name}', "
-                            f"received invalid type:\n{loss_weights[name]}"
-                        )
-                for name in output_names:
-                    if name in loss_weights:
-                        flat_loss_weights.append(loss_weights[name])
-                    else:
-                        flat_loss_weights.append(1.0)
-            else:
-                flat_loss_weights = [1.0 for _ in flat_losses]
+                filtered_y_true_keys.extend(list(y_true_keys - loss_keys))
+        filtered_y_pred_keys = set(filtered_y_pred_keys)
+        filtered_y_true_keys = set(filtered_y_true_keys)
+
+        flat_losses = tree.flatten(loss)
+        if loss_weights is None:
+            flat_loss_weights = [None] * len(flat_losses)
+        else:
+            flat_loss_weights = tree.flatten(loss_weights)
+            for loss_weight in flat_loss_weights:
+                if not isinstance(loss_weight, (int, float, type(None))):
+                    raise TypeError(
+                        "When providing the `loss_weights` argument, each "
+                        "element should be a Python int, float (the weighting "
+                        "coefficient corresponding to the loss for that "
+                        "output) or `None`."
+                        f"Received: loss_weights={loss_weights}"
+                    )
+            if len(flat_loss_weights) != len(flat_losses):
+                raise ValueError(
+                    "When providing the `loss_weights` argument, it should "
+                    "have equal length of `loss` argument. "
+                    f"Received: loss_weights length={len(flat_loss_weights)}, "
+                    f"loss legnth={len(flat_losses)}"
+                )
+        flat_y_true_specs, flat_y_pred_specs = self._get_specs(flat_losses)
```

Comment:
I don't quite follow how this is used. Could it be simplified?

Reply:
I have added some comments to clarify the entire procedure.
Essentially, we need `y_true_spec` and `y_pred_spec` in a flat form because we have flattened the losses in that manner.
In `call`, these specs are necessary to correctly pack `y_true` and `y_pred` before iterating through them with `self.flat_losses`.